### DISCUSSION TITLE
Translate File Field

### DISCUSSION THREAD

**Discussion by pierreberchtold** [spatie/nova-translatable#60] [Q&A]
When I try to translate a Nova File Field I get error when saving. The File Field is trying to save the `translations_FIELDNAME_LANG` attribute and not the `FIELDNAME` attribute.

I don't really see through how `nova-translatable` and the File Field are working so I have no idea what the best approach would be to fix this. If you have a hint I'd be happy to do a PR.

**Comment by NSpehler** [comment]
Same issue here, I'm getting an error `Call to a member function store() on null` when using Translatable on a File field.

**Comment by danyelkeddah** [comment]
Hey everyone, Any updates ?
well I have the same issue with Trix field too, after dropping an image inside the field it is trying to upload it and store its data into the table on this end point
`nova-api/{resource}/trix-attachment/translations_FIELDNAME_LANG`
 for some reason it's dumping `404 (Not Found)` to the browser console

**Comment by n0n0n0n0** [comment]
same error, looking for solution!

**Comment by tonci** [comment]
As a workaround, just add:

`Trix::make('translations_FIELDNAME_LANG')->withFiles('public')->hideWhenCreating()->hideWhenUpdating()->hideFromDetail()`

so your fields function will look like:
```
public function fields(Request $request)
    {
        return [
            ID::make()->sortable(),
            \Spatie\NovaTranslatable\Translatable::make([
                Trix::make('content')->withFiles('s3'),
            ]),
            Trix::make('translations_content_en')->withFiles('s3')->hideWhenCreating()->hideWhenUpdating()->hideFromDetail(),
            Trix::make('translations_content_dk')->withFiles('s3')->hideWhenCreating()->hideWhenUpdating()->hideFromDetail(),
        ];
    }
```
This needs to be done for each language you want to support (yeah... ugly).
Tested this with amazon s3 and seem to also work when you delete the files from the trix field.

**Comment by jrmadsen67** [comment]
when trying @tonci method I get:

![image](https://user-images.githubusercontent.com/821958/56167872-60228c00-6014-11e9-8fbd-3109f7a1b4d7.png)

Trix images work fine without the wrapper. 
If I have `Trix::make('body')->withFiles('public'),` shouldn't my workaround field be:
`Trix::make('translations_body_de-AT')->withFiles('public')->hideWhenCreating()->hideWhenUpdating()->hideFromDetail(),`?

**Comment by mkhmylife** [comment]
@jrmadsen67 You have to use `Trix::make('translations_body_de-AT', 'translations_body_de-AT')->withFiles('public')->hideWhenCreating()->hideWhenUpdating()->hideFromDetail()` to make it work

**Comment by ilikourou** [comment]
Is there any update here?

Both Trix and Froala fields don't work with spatie/nova-translatable when it comes to file uploads/attachments.

**Comment by Grunkhead** [comment]
Same issue here. Seems like the call is of the wrong type.

![image](https://user-images.githubusercontent.com/746279/62119147-3b9aa900-b2bf-11e9-9663-8245afc0d3d6.png)

**Comment by Grunkhead** [comment]
> As a workaround, just add:
> 
> `Trix::make('translations_FIELDNAME_LANG')->withFiles('public')->hideWhenCreating()->hideWhenUpdating()->hideFromDetail()`
> 
> so your fields function will look like:
> 
> ```
> public function fields(Request $request)
>     {
>         return [
>             ID::make()->sortable(),
>             \Spatie\NovaTranslatable\Translatable::make([
>                 Trix::make('content')->withFiles('s3'),
>             ]),
>             Trix::make('translations_content_en')->withFiles('s3')->hideWhenCreating()->hideWhenUpdating()->hideFromDetail(),
>             Trix::make('translations_content_dk')->withFiles('s3')->hideWhenCreating()->hideWhenUpdating()->hideFromDetail(),
>         ];
>     }
> ```
> 
> This needs to be done for each language you want to support (yeah... ugly).
> Tested this with amazon s3 and seem to also work when you delete the files from the trix field.

Do not forget to place it in the main array within the fields method of het Resource. It acts like a hidden field.

Just to be sure. Thanks!

**Comment by ilikourou** [comment]
Keep in mind that this workaround seems to be working only with Trix field type.

If you are using a different type of editor (eg Froala), this is not enough to make it work, unfortunately.

**Comment by Naoray** [comment]
Had the same problem with file uploads using the `Trix` field. Fixed it by creating my own `Translatable` field by extending Spatie's and overriding the `createTranslationFields()` method.

```php
// User Resource
public function fields(Request $request)
    {
        return [
              //...

              Translatable::make([
                  Trix::make('About')
                       ->withFiles('public')
                       ->alwaysShow(),
                  ])
        ];
    }

// Translatable field
namespace App\Nova\Fields;

use Laravel\Nova\Fields\Trix;
use Laravel\Nova\Fields\Field;
use Spatie\NovaTranslatable\Translatable as BaseTranslatable;

class Translatable extends BaseTranslatable
{
    protected function createTranslatableFields()
    {
        parent::createTranslatableFields();

        if ($this->onIndexPage()) {
            return;
        }

        collect($this->locales)
            ->crossJoin($this->originalFields)
            ->eachSpread(function (string $locale, Field $field) {
                if ($field instanceof Trix && $field->withFiles) {
                    $this->data[] = get_class($field)::make('translations_' . $field->attribute . '_' . $locale)
                        ->onlyOnIndex()
                        ->showOnIndex(false)
                        ->withFiles($field->disk, $field->storagePath);
                }
            });
    }
}
```

@freekmurze I don't feel that this solution is clean in any way, but maybe it sparks an idea on your side on how to solve this in a more elegant way? Otherwise I am also open to create a PR with my fix if you want.

**Comment by JoJuhasz** [comment]
What about File fields ?
Is this not the subject  of this issue ? This is not about Trix

**Comment by yassinetrbls** [comment]
Any news ?

**Comment by mrtawil** [comment]
+1

**Comment by FelipeKraefft** [comment]
News?
